src/target/riscv/riscv.c

```diff
@@ -2035,7 +2046,15 @@ int riscv_get_register_on_hart(struct target *target, riscv_reg_t *value,
 		int hartid, enum gdb_regno regid)
 {
 	RISCV_INFO(r);
+
+	if (hartid != riscv_current_hartid(target))
+		riscv_invalidate_register_cache(target);
+
 	int result = r->get_register(target, value, hartid, regid);
+
+	if (hartid != riscv_current_hartid(target))
```

Comment:
I'm curious how the current hartid can not match the hartid of the last hart we read a register from.

Reply:
My commit says "The 0.13 code now caches registers interally, so when reading registers on a different hart we need to invalidate the cache.", though I'm not sure how that's relevant.

Reply:
Let's leave it for now. It should be a harmless check.